File: kreuzberg/packages/java/src/test/java/dev/kreuzberg/SerializationTest.java
package dev.kreuzberg;

import static org.junit.jupiter.api.Assertions.*;

import com.fasterxml.jackson.databind.ObjectMapper;
import dev.kreuzberg.config.ExtractionConfig;
import dev.kreuzberg.config.OcrConfig;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

/**
 * Cross-language serialization tests for Java bindings.
 *
 * Validates that ExtractionConfig serializes consistently with other language
 * bindings.
 */
@DisplayName("ExtractionConfig Serialization")
class SerializationTest {

	private ObjectMapper objectMapper;

	@BeforeEach
	void setUp() {
		objectMapper = new ObjectMapper();
	}

	@Test
	@DisplayName("should serialize minimal config to JSON")
	void testMinimalSerialization() throws Exception {
		ExtractionConfig config = ExtractionConfig.builder().build();
		String json = config.toJson();

		assertNotNull(json);
		assertTrue(json.contains("use_cache"));
		assertTrue(json.contains("enable_quality_processing"));
		assertTrue(json.contains("force_ocr"));
	}

	@Test
	@DisplayName("should serialize config with custom values")
	void testCustomValuesSerialization() throws Exception {
		ExtractionConfig config = ExtractionConfig.builder().useCache(true).enableQualityProcessing(false)
				.forceOcr(true).build();

		String json = config.toJson();

		ExtractionConfig restored = ExtractionConfig.fromJson(json);

		assertEquals(true, restored.isUseCache());
		assertEquals(false, restored.isEnableQualityProcessing());
		assertEquals(true, restored.isForceOcr());
	}

	@Test
	@DisplayName("should preserve field values after serialization")
	void testFieldPreservation() throws Exception {
		ExtractionConfig original = ExtractionConfig.builder().useCache(false).enableQualityProcessing(true).build();

		String json = original.toJson();
		ExtractionConfig restored = ExtractionConfig.fromJson(json);

		assertEquals(original.isUseCache(), restored.isUseCache());
		assertEquals(original.isEnableQualityProcessing(), restored.isEnableQualityProcessing());
	}

	@Test
	@DisplayName("should handle round-trip serialization")
	void testRoundTripSerialization() throws Exception {
		ExtractionConfig config1 = ExtractionConfig.builder().useCache(true).enableQualityProcessing(false).build();

		String json1 = config1.toJson();
		ExtractionConfig config2 = ExtractionConfig.fromJson(json1);
		String json2 = config2.toJson();

		// Should produce equivalent JSON
		ExtractionConfig parsed1 = ExtractionConfig.fromJson(json1);
		ExtractionConfig parsed2 = ExtractionConfig.fromJson(json2);

		assertEquals(parsed1.isUseCache(), parsed2.isUseCache());
		assertEquals(parsed1.isEnableQualityProcessing(), parsed2.isEnableQualityProcessing());
		assertEquals(parsed1.isForceOcr(), parsed2.isForceOcr());
	}

	@Test
	@DisplayName("should use snake_case field names in JSON")
	void testSnakeCaseFieldNames() throws Exception {
		ExtractionConfig config = ExtractionConfig.builder().build();
		String json = config.toJson();

		assertTrue(json.contains("use_cache"));
		assertTrue(json.contains("enable_quality_processing"));
		assertTrue(json.contains("force_ocr"));
	}

	@Test
	@DisplayName("should serialize nested OCR config")
	void testNestedOcrConfig() throws Exception {
		OcrConfig ocrConfig = OcrConfig.builder().backend("tesseract").language("eng").build();

		ExtractionConfig config = ExtractionConfig.builder().ocr(ocrConfig).build();

		String json = config.toJson();

		assertTrue(json.contains("\"ocr\""));
		assertTrue(json.contains("\"backend\""));
		assertTrue(json.contains("\"tesseract\""));
		assertTrue(json.contains("\"language\""));
		assertTrue(json.contains("\"eng\""));
	}

	@Test
	@DisplayName("should handle null values correctly")
	void testNullValueHandling() throws Exception {
		ExtractionConfig config = ExtractionConfig.builder().build();

		String json = config.toJson();
		ExtractionConfig restored = ExtractionConfig.fromJson(json);

		assertNull(restored.getOcr());
		assertNull(restored.getChunking());
	}

	@Test
	@DisplayName("should maintain stability during serialization")
	void testStabilityDuringSerialization() throws Exception {
		ExtractionConfig config = ExtractionConfig.builder().useCache(true).build();

		String json1 = config.toJson();
		String json2 = config.toJson();
		String json3 = config.toJson();

		assertEquals(json1, json2);
		assertEquals(json2, json3);
	}

	@Test
	@DisplayName("should serialize all mandatory fields")
	void testMandatoryFields() throws Exception {
		ExtractionConfig config = ExtractionConfig.builder().build();
		String json = config.toJson();

		ExtractionConfig parsed = ExtractionConfig.fromJson(json);

		assertNotNull(parsed.isUseCache());
		assertNotNull(parsed.isEnableQualityProcessing());
		assertNotNull(parsed.isForceOcr());
	}

	@Test
	@DisplayName("should deserialize from JSON string")
	void testDeserialization() throws Exception {
		String json = "{\"use_cache\":true,\"enable_quality_processing\":false,\"force_ocr\":true}";

		ExtractionConfig config = ExtractionConfig.fromJson(json);

		assertEquals(true, config.isUseCache());
		assertEquals(false, config.isEnableQualityProcessing());
		assertEquals(true, config.isForceOcr());
	}
}

File: kreuzberg/packages/csharp/SerializationTest.cs
using System;
using System.Text.Json;
using Xunit;
using Kreuzberg;

namespace Kreuzberg.Tests
{
    /// <summary>
    /// Cross-language serialization tests for C# bindings.
    ///
    /// Validates that ExtractionConfig serializes consistently with other language bindings.
    /// </summary>
    public class SerializationTest
    {
        private readonly JsonSerializerOptions _jsonOptions = new()
        {
            PropertyNamingPolicy = JsonNamingPolicy.CamelCase,
            WriteIndented = false
        };

        [Fact(DisplayName = "Should serialize minimal config to JSON")]
        public void TestMinimalSerialization()
        {
            var config = new ExtractionConfig();
            var json = JsonSerializer.Serialize(config, _jsonOptions);

            Assert.NotNull(json);
            Assert.Contains("useCache", json);
            Assert.Contains("enableQualityProcessing", json);
            Assert.Contains("forceOcr", json);
        }

        [Fact(DisplayName = "Should serialize config with custom values")]
        public void TestCustomValuesSerialization()
        {
            var config = new ExtractionConfig
            {
                UseCache = true,
                EnableQualityProcessing = false,
                ForceOcr = true
            };

            var json = JsonSerializer.Serialize(config, _jsonOptions);
            var restored = JsonSerializer.Deserialize<ExtractionConfig>(json, _jsonOptions);

            Assert.NotNull(restored);
            Assert.True(restored.UseCache);
            Assert.False(restored.EnableQualityProcessing);
            Assert.True(restored.ForceOcr);
        }

        [Fact(DisplayName = "Should preserve field values after serialization")]
        public void TestFieldPreservation()
        {
            var original = new ExtractionConfig
            {
                UseCache = false,
                EnableQualityProcessing = true
            };

            var json = JsonSerializer.Serialize(original, _jsonOptions);
            var restored = JsonSerializer.Deserialize<ExtractionConfig>(json, _jsonOptions);

            Assert.NotNull(restored);
            Assert.Equal(original.UseCache, restored.UseCache);
            Assert.Equal(original.EnableQualityProcessing, restored.EnableQualityProcessing);
        }

        [Fact(DisplayName = "Should handle round-trip serialization")]
        public void TestRoundTripSerialization()
        {
            var config1 = new ExtractionConfig
            {
                UseCache = true,
                EnableQualityProcessing = false
            };

            var json1 = JsonSerializer.Serialize(config1, _jsonOptions);
            var config2 = JsonSerializer.Deserialize<ExtractionConfig>(json1, _jsonOptions);
            var json2 = JsonSerializer.Serialize(config2, _jsonOptions);

            // Parse both JSONs and compare
            var element1 = JsonDocument.Parse(json1).RootElement;
            var element2 = JsonDocument.Parse(json2).RootElement;

            Assert.Equal(element1.GetRawText(), element2.GetRawText());
        }

        [Fact(DisplayName = "Should use camelCase field names")]
        public void TestCamelCaseFieldNames()
        {
            var config = new ExtractionConfig { UseCache = true };
            var json = JsonSerializer.Serialize(config, _jsonOptions);

            Assert.Contains("useCache", json);
            Assert.Contains("enableQualityProcessing", json);
            Assert.Contains("forceOcr", json);

            Assert.DoesNotContain("use_cache", json);
            Assert.DoesNotContain("enable_quality_processing", json);
            Assert.DoesNotContain("force_ocr", json);
        }

        [Fact(DisplayName = "Should serialize nested OCR config")]
        public void TestNestedOcrConfig()
        {
            var config = new ExtractionConfig
            {
                Ocr = new OcrConfig
                {
                    Backend = "tesseract",
                    Language = "eng"
                }
            };

            var json = JsonSerializer.Serialize(config, _jsonOptions);

            Assert.Contains("ocr", json.ToLower());
            Assert.Contains("tesseract", json);
            Assert.Contains("eng", json);
        }

        [Fact(DisplayName = "Should handle null values correctly")]
        public void TestNullValueHandling()
        {
            var config = new ExtractionConfig
            {
                Ocr = null,
                Chunking = null
            };

            var json = JsonSerializer.Serialize(config, _jsonOptions);
            var restored = JsonSerializer.Deserialize<ExtractionConfig>(json, _jsonOptions);

            Assert.NotNull(restored);
            Assert.Null(restored.Ocr);
            Assert.Null(restored.Chunking);
        }

        [Fact(DisplayName = "Should maintain immutability during serialization")]
        public void TestImmutabilityDuringSerialization()
        {
            var config = new ExtractionConfig { UseCache = true };

            var json1 = JsonSerializer.Serialize(config, _jsonOptions);
            var json2 = JsonSerializer.Serialize(config, _jsonOptions);
            var json3 = JsonSerializer.Serialize(config, _jsonOptions);

            Assert.Equal(json1, json2);
            Assert.Equal(json2, json3);
        }

        [Fact(DisplayName = "Should serialize all mandatory fields")]
        public void TestMandatoryFields()
        {
            var config = new ExtractionConfig();
            var json = JsonSerializer.Serialize(config, _jsonOptions);
            var element = JsonDocument.Parse(json).RootElement;

            Assert.True(element.TryGetProperty("useCache", out _), "useCache field is missing");
            Assert.True(element.TryGetProperty("enableQualityProcessing", out _), "enableQualityProcessing field is missing");
            Assert.True(element.TryGetProperty("forceOcr", out _), "forceOcr field is missing");
        }

        [Fact(DisplayName = "Should deserialize from JSON string")]
        public void TestDeserialization()
        {
            var json = """{"useCache":true,"enableQualityProcessing":false,"forceOcr":true}""";
            var config = JsonSerializer.Deserialize<ExtractionConfig>(json, _jsonOptions);

            Assert.NotNull(config);
            Assert.True(config.UseCache);
            Assert.False(config.EnableQualityProcessing);
            Assert.True(config.ForceOcr);
        }

        [Fact(DisplayName = "Should produce valid JSON")]
        public void TestValidJsonOutput()
        {
            var config = new ExtractionConfig { UseCache = true };
            var json = JsonSerializer.Serialize(config, _jsonOptions);

            // Should not throw
            var _ = JsonDocument.Parse(json);
            Assert.NotEmpty(json);
        }

        [Fact(DisplayName = "Should pretty-print JSON")]
        public void TestPrettyPrint()
        {
            var config = new ExtractionConfig { UseCache = true };
            var prettyOptions = new JsonSerializerOptions
            {
                PropertyNamingPolicy = JsonNamingPolicy.CamelCase,
                WriteIndented = true
            };

            var json = JsonSerializer.Serialize(config, prettyOptions);

            // Pretty JSON should have newlines
            Assert.Contains("\n", json);

            // Should still be valid JSON
            var _ = JsonDocument.Parse(json);
        }
    }
}
